Add the avifDecoder option 'imageDimensionLimit'

wantehchang opened the pull request:
Add the avifDecoder option 'imageDimensionLimit', default to 32768. This
allows the decoder to reject large image widths or heights by default.

Add the internal utility function avifDimensionsTooLarge(), declared in
include/avif/internal.h and defined in src/avif.c.

Bug: b/242362974

joedrago reviewed (approved):
I like this change. If we ever want to impose additional limits in the future, we should put all of the limits into a struct and pass that around instead.

jzern commented on include/avif/avif.h:
```diff
@@ -878,11 +881,16 @@ typedef struct avifDecoder
 
+    // This represents the maximum dimension of a image (width or height) that libavif should
+    // attempt to decode. It defaults to AVIF_DEFAULT_IMAGE_DIMENSION_LIMIT, and can be set to a
+    // a smaller value. Set it to 0 to ignore the limit.
```
extra 'a' at the end of line 890

wantehchang replied:
Done.

wantehchang commented on include/avif/avif.h:
```diff
@@ -878,11 +881,16 @@ typedef struct avifDecoder
     // Note: Only some underlying AV1 codecs support a configurable size limit (such as dav1d).
     uint32_t imageSizeLimit;
 
+    // This represents the maximum dimension of a image (width or height) that libavif should
```
Done.

wantehchang commented on apps/avifdec.c:
```diff
@@ -45,6 +45,9 @@ static void syntax(void)
     printf("    --size-limit C    : Specifies the image size limit (in total pixels) that should be tolerated.\n");
     printf("                        Default: %u, set to a smaller value to further restrict.\n", AVIF_DEFAULT_IMAGE_SIZE_LIMIT);
+    printf("  --dimension-limit C : Specifies the image dimension limit (width or height) that should be tolerated.\n");
+    printf("                        Default: %u, set to a smaller value to further restrict, or 0 to ignore.\n",
```
I just noticed a copy-and-paste error here. "set to a smaller value to further restrict" is correct for `--size-limit` because we don't allow `--size-limit` to be set to a value larger than the default value. But there is no such constraint for `--dimension-limit`. I will change this line to:

```
    printf("                        Default: %u, set to 0 to ignore.\n",
```

wantehchang commented on include/avif/avif.h:
```diff
@@ -877,12 +880,17 @@ typedef struct avifDecoder
 
+    // This represents the maximum dimension of an image (width or height) that libavif should
+    // attempt to decode. It defaults to AVIF_DEFAULT_IMAGE_DIMENSION_LIMIT, and can be set to a
+    // smaller value. Set it to 0 to ignore the limit.
```
Similarly, "and can be set to a smaller value" is a copy-and-paste error. There is no such constraint for `imageDimensionLimit`. I will remove it.

wantehchang commented:
Please review again. The only changes are:

1. Described this change in CHANGELOG.md.
2. Modified the help message for `--dimension-limit` in avifdec.c.
3. Modified the comment for the `imageDimensionLimit` field in avif.h.

Thank you.
